### PR TITLE
support local indexing

### LINKED ISSUE #23: bootstrap index for download urls
many popular packages do not host their documentation on read the docs, nor do they list a Documentation project_url in their PyPI project metadata. We can ship with some defaults to brighten up the user experience.

### PR DESCRIPTION
**Pull request by jpurviance**
closes #23

### CODE REVIEW

File: attending/index.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import csv
+from pathlib import Path
+import os
+
+
+def load_index(file: Path):
+    project_index = {}
+    assert file.is_file()
```
Comment (ivanov) [inline attending/index.py:8]:
```suggestion
```

Also, assertions get eliminate when running python with optimization turned on.

```
$ python -c "assert True == False, 'oops'; print('ok')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AssertionError: oops
$ python -O -c "assert True == False, 'oops'; print('ok')"
ok
```

It's no better to fail here compared to failing on the file `open()` call, but leaving this out allows us to accept strings and bytes as the input parameter.

File: attending/index.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import csv
+from pathlib import Path
+import os
+
+
+def load_index(file: Path):
```
Comment (ivanov) [inline attending/index.py:6]: so I know you like Paths, but once you eliminate the assertion check below, there's no reason to limit `file` to a Path, strings are fine here, too. See [the PEP 519 specification](https://www.python.org/dev/peps/pep-0519/) which suggests the path annotation of `path: Union[PathLike, str, bytes]` for the first parameter of os.fspath, for example.

Also, calling a variable `file` is <strike>poor practice for shadowing ....</strike> oops, I forgot that Python 3 got rid of the `file` builtin... it still looks weird to me... but that probably more of a "who moved my cheese" resistance that I need to work through.

```suggestion
def load_urls(file):
```

File: attending/library.py [hunk 3]
```diff
@@ -159,6 +160,13 @@ def fetch_via_name(module, version=None, url=None):
                 raise options_exhausted(module, version)
         else:
             return fetch_via_module(python_module, version)
-    
     return lib.fetch(module, version, url)
 
+
+def fetch_via_local_index(module, home=Path().home()):
```
Comment (ivanov) [inline attending/library.py:166]: Let's keep the interface of the other fetch modules
```suggestion
def fetch_via_local_index(module, version=None, home=Path().home()):
```
and specify a placeholder for where the version portion of the string goes in the URL. That way we don't hardcode the urls in the CSV to a particular version.

File: attending/library.py [hunk 4]
```diff
@@ -159,6 +160,13 @@ def fetch_via_name(module, version=None, url=None):
...
+    index = load_attending_index()
+    local_index_file = Path().home() / ".attending" / "index.csv"
```
Comment (ivanov) [inline attending/library.py:168]:
```suggestion
    local_index_file = home / ".attending" / "index.csv"
```

File: attending/index.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    project_index = {}
+    assert file.is_file()
+    with open(str(file), 'r') as index:
+        index_reader = csv.reader(index)
+        for row in index_reader:
```
Comment (ivanov) [inline attending/index.py:11]: How about unpacking this as 
```suggestion
        for module,project,url in url_reader:
            project_url[module] = url
            project_url[project] = url
```

File: attending/index.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import csv
+from pathlib import Path
+import os
+
+
+def load_index(file: Path):
+    project_index = {}
+    assert file.is_file()
+    with open(str(file), 'r') as index:
```
Comment (ivanov) [inline attending/index.py:9]: no need for `str` here, and 'r' is the default mode anyway.
```suggestion
    with open(file) as urls:
```

**Review by ivanov — CHANGES_REQUESTED** [review]
Looks good over all. Can we change the 'index' name to "supplemental" - as in, this will be our supplemental insurance policy for modules that are not on ReadTheDocs and that have not adopted __doc_url__ yet? I'm open to any other name, I just do not want to reserve "index" for indexing the contents of the docs, or at least that's what it invokes in my head, as opposed to being a list of doc urls.

`doc_urls.csv` would also work, so maybe `s/index/doc_urls/g` would do the trick? How does that sound? or even just "urls.csv" since the file is already in the `.attending` folder, so there's no ambiguity about what it contains. 

Upon reflection, `supplemental` is too long of a name and too cutesy, let's just go with `urls`... Some of my suggestions contain this, but don't accept them through this interface, since I didn't go through replacing all of the necessary places where changes are needed.

File: attending/library.py [hunk 7]
```diff
@@ -77,17 +78,17 @@ def get_edition(self, name, version):
         return Module(self.location, name)[version]
 
 
-def attending_doc(module, version=None):
+def attending_doc(module, version=None, home=Path().home()):
```
Comment (scopatz) [inline attending/library.py:81]: I forget, are Path objects mutable?

Comment (ivanov) [reply attending/library.py:81]: Paths are immutable and hashable (so you can use them as keys in a dictionary, for example). https://docs.python.org/3/library/pathlib.html#general-properties

**Pull request merged**